modules/cudastereo/test/test_stereo.cpp

```diff
@@ -209,6 +209,112 @@ INSTANTIATE_TEST_CASE_P(CUDA_Stereo, ReprojectImageTo3D, testing::Combine(
     testing::Values(MatDepth(CV_8U), MatDepth(CV_16S)),
     WHOLE_SUBMAT));
 
+////////////////////////////////////////////////////////////////////////////////
+// StereoSGM
+
+struct StereoSGM_Test_Params
+{
+    String leftSrcImagePath;
+    String rightSrcImagePath;
+    String leftFixedImagePath;
+    int dispScaleFactor;
+    int dispUnknVal;
+};
+
+std::ostream& operator<<(std::ostream& os, const StereoSGM_Test_Params& params)
+{
+    os << "(" << params.leftSrcImagePath << ", " << params.rightSrcImagePath << ") - > " << params.leftFixedImagePath;
+    os << ", scale factor = " << params.dispScaleFactor << ", unknown value = " << params.dispUnknVal;
+    return os;
+}
+
+PARAM_TEST_CASE(StereoSGM, cv::cuda::DeviceInfo, StereoSGM_Test_Params)
+{
+    cv::cuda::DeviceInfo devInfo;
+    StereoSGM_Test_Params params;
+
+    virtual void SetUp()
+    {
+        devInfo = GET_PARAM(0);
+        params = GET_PARAM(1);
+
+        cv::cuda::setDevice(devInfo.deviceID());
+    }
+};
+
+CUDA_TEST_P(StereoSGM, Regression)
+{
+    Mat h_leftImage = imread(params.leftSrcImagePath, IMREAD_GRAYSCALE);
+    Mat h_rightImage = imread(params.rightSrcImagePath, IMREAD_GRAYSCALE);
+    Mat trueLeftDisp = imread(params.leftFixedImagePath, IMREAD_GRAYSCALE);
+    assert(!h_leftImage.empty() && !h_rightImage.empty() && !trueLeftDisp.empty());
+
+    Mat tmp;
+    trueLeftDisp.convertTo(tmp, CV_32FC1, 1.f / params.dispScaleFactor);
+    trueLeftDisp = tmp;
+    tmp.release();
+
+    GpuMat d_leftImage, d_rightImage, d_leftDisp;
+    d_leftImage.upload(h_leftImage);
+    d_rightImage.upload(h_rightImage);
+
+    cv::Ptr<cv::cuda::StereoSGM> sgm = cv::cuda::createStereoSGM(64, 10, 120, 5);
+    sgm->compute(d_leftImage, d_rightImage, d_leftDisp);
+
+    Mat h_leftDisp;
+    d_leftDisp.download(h_leftDisp);
+    h_leftDisp.convertTo(tmp, CV_32FC1, 1.f / StereoMatcher::DISP_SCALE);
+    h_leftDisp = tmp;
+    tmp.release();
+
+    Mat unknMask;
+    absdiff(trueLeftDisp, Scalar(params.dispUnknVal), unknMask);
+    unknMask = unknMask < std::numeric_limits<float>::epsilon();
+
+    double allRMS = 1.0 / sqrt(trueLeftDisp.cols * trueLeftDisp.rows) * cvtest::norm(h_leftDisp, trueLeftDisp, NORM_L2, unknMask);
+    EXPECT_LT(allRMS, 5);
```

Comment:
I would like to compare each previous value like
https://github.com/opencv/opencv_extra/blob/c721e8203239823c7dc5a3f7579dde23799c5e8a/testdata/cv/stereomatching/algorithms/stereosgbm_res.xml

Reply:
Done